compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -130,7 +125,8 @@ pub fn intern_const_alloc_recursive<
 
     // Initialize recursive interning.
     let base_alloc_id = ret.ptr().provenance.unwrap().alloc_id();
-    let mut todo = vec![(base_alloc_id, base_mutability)];
+    let mut todo: Vec<_> =
+        intern_shallow(ecx, base_alloc_id, base_mutability).unwrap().map(|prov| prov).collect();
```

Comment:
I don't understand why you want to do this before the loop. Why the code duplication?

Reply:
We duplicate other code though. I was quite happy with having everything inside that one loop. It's rather unusual in recursive traversals to have the first round outside the loop in this way, so it makes the code harder to follow, too.

But I'm also okay changing it again for now, and once I understood https://github.com/rust-lang/rust/pull/116564 maybe I can see if I can bring the more uniform structure back.

Reply:
To me it is easier to follow, because there's no runtime decision on the mutability anymore. But yea, we can discuss this once I have even more code that supports doing it this way